Availability Labels to Shift Users

amanmibra opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Implemented the availability functionality to adding users to a shift when creating or updating a shift. The availability of a user is calculated by their lowest possible availability within a range of a starting and ending time of a shift. 

Each level of availability is represented by the colors: Red (Unavailable), Yellow (Somewhat), and Green (Available). 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This is a continuation of #39 and it is actually implementing the information that users give to their team about their weekly availability to the process of creating a shift.

## How Has This Been Tested?
**Remember:** _Always migrate your database in case of changes (yarn dev:db:migrate)_
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
1. New seeds have been created to give other users randomized availabilities, so do the following:
   a. Run `yarn flush`, and respond 'y'
   b. Good to go! Run 'yarn dev:start`
2. Create a user, and join a team.
3. Create a shift, and add users. See that the dropdown now has users aprox. availability labeled.
4. Update a shift, and add/remove users. Check the dropdown.

## Screenshots (if appropriate):
![image](https://user-images.githubusercontent.com/13587407/50426501-2eb36400-085d-11e9-8c6e-d85c42633ee6.png)


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

amanmibra committed:
Built API endpoint to find user aprox availability for a new shift

amanmibra committed:
Translating JS date objects to availability indicies

amanmibra committed:
Added availability feature to CreateShiftForm

amanmibra committed:
Moved getShiftAvail util

amanmibra committed:
Added availability feature to UpdateShiftForm

amanmibra committed:
Added db migrate script

vinitbytheminute reviewed (dismissed):
### **Overview**
Great job in the initial implementation. The availability each person implemented during the login portion is properly shown when creating an event. For instance, if I said I was available from 2 to 7 pm, that availability is shown with a green dot when I am in the calendar portion creating a shift between 2 to 7 pm right next to my name. Therefore, the intail implementation is great. I also did not catch any bugs with the code. However, after playing with this new implementation, I believe additions are needed in order to make this availability feature truly useful. 

**Additions**
I mentioned this in a private message to you. I think it is important to implement the range when each member is available for two main reasons. The first reason is to make it much easier for the captain to know when someone is available relative to the time slot they had chosen. For instance, if I said I was available from 2 to 5pm, but the captain is making a shift between 130pm to 3pm, I would show up unavailable because I am not available between 130 to 2pm. This is unfortunate because I am actually available during portions of that time (from 2 to 3pm and available after from 3 to 5pm). The captain and members should know this. Here is a [link](https://duke.box.com/s/6cvbasy1bclea46dppjzudsr6slowbhl) of a video of me and usertester explaining the importance of the feature I described above along with a visual of how to display this information. 

Second reason why it is important to show a range is to make it easier for a member to look at who else may be available at a certain time in case she needs to switch his/her shift. For instance, if I have a shift on Tuesday from 2:30 to 6pm, but I now can only be there at 3:30 the earliest, I want to be able to press on my shift without needing to create a new shift to see who else is potentially available from 2:30 to 4pm. What if "Max" is available from 1 to 4. With the current implementation, it will not show up that Max is free from 1 to 3 since he is unavailable from  4 to 6pm within the current shift time period between (1 to 6pm). Therefore, I wont know that Max would actually be an optimal person to split my shift with and contact.

amanmibra commented:
@vinitbytheminute 

#### Response to "Night Shift Availability Bugs"
That's not a bug. This is just a misunderstanding that comes with replacing hours with the words "Night Shift" and how the availability component is designed. On the grid, night shift comes at the end of a column, so, the night shift of a day is then left to be assumed to be of the night of that day, rather than the morning of, which is the assumption you are making. 

You would have easily realized this if you checked Wednesday nights. We already decided as team that a day as tenter functions from "7AM - 2AM", and then we have night. So, a being at the tent at 1AM Thursday morning is a part of a Wednesday shift, and sleeping in the tent on a Monday morning is a part of the Sunday night shift. If this confuses you, then this call for a redesign on the availability not the availability algorithm, as it only follows the design.

#### Response to "Additions"
I already told you we are not doing this now.

mafuvadze commented:
Thank you, Aman. Very cool. 

- I think the colors for the availability are fairly intuitive,  but I think they might confuse some users. I suggest you add a key at the bottom that states that green=available yellow=limited availability red=busy.

- This isn't absolutely necessary, but I think it would helpful if we had the number of hours that person has done in parenthesis. ex Aman Ibrahim (12 hrs this week). It may help the captain assign.

- Not clear that you have to double click to get to the available state. Maybe add a note to the availability screen to let users know?

amanmibra commented:
@mafuvadze 
- I can def. add a key for this. 
- I will be building a hour counter alongside Vinit's dashboard, once it is pushed to `dashboard-hours-amanmibra`, so, we can possibly implement that later into the dropdown.
- I can see what's the best way of relaying that message to users. Good idea.

Abdulla-Shahid commented:
Awesome job Aman!
I thought the UI looked great and did exactly what we wanted it to do! I had some minor edits/suggestions that I thought you could look into if you had time, but aren't essential:   

- I think it would be more useful/save time if a person was able to switch from unavailability to availability in one click rather than two clicks. I feel like most people will either select available vs not available rather than available, somewhat, not-available, so the order of clicks should be: 1= available, 2 = somewhat-available, and then 3 would take you back to normal state of being not available. Also, some information in the key or a separate instructions card above or below the key about how changing availability works could be very helpful for the user. 

- I'm not sure how difficult this will be, but would it be possible to have a drag/highlight option of some sort to select all the dates someone's available rather than individually pressing all the dates? Or even just a text/dialog box above or below the calendar where a user could just select the day of the week and then enter in the continuous times there free and have it automatically checked off? Ex: Monday: 5am - 7pm mark available. Not a necessity by any means but would be cool to implement now or in the future I think. 

**Also on a side note, when a user is using an email that's already in use, you're still allowed to progress/use that email. When I was on the calendar page the error message was still there: 
![image](https://user-images.githubusercontent.com/22209648/50468663-1899c680-0977-11e9-8ae1-9798d61d720e.png)

Not sure if that was something anyone's worked on yet, but I thought it would be good to point out so everyone in the team could see it. 

 Let me know if you need me to clarify anything! Again, amazing job Aman!

amanmibra committed:
Made the following changes to Availability:
- Key Redesign
- Changed location of time rows
- Removed lodash
- Updated API requests

amanmibra committed:
Added popup for color info about user dropdown

amanmibra committed:
Added passcode to TeamProfile info

amanmibra commented:
@mafuvadze 
Check out my recent changes. 

1. Go to whichever Availability component (whether in sign up or user profile) and see my redesign to the key, and let me know what you think of the instructions given to the users. My goal is to keep as short and straightfoward as possible.
2. Go and create or update a shift, you can see that there is a hover info icon that gives a key for what the colors mean without being intrusive of the original design of the dropdown.

Also, I noticed you removed it in your comment, but I also added the passcode to the Team Profile.

amanmibra commented on app/javascript/src/components/availability/index.jsx:
```diff
@@ -3,30 +3,27 @@ import React, { Component } from 'react';
 import { Grid, Segment, Divider, Button, Icon, Card } from 'semantic-ui-react';
 
-// helpers
-import _ from 'lodash';
```
@rikkikendall Please do not use lodash. Most, if not all of it's functionality, can be found on base JS.

amanmibra committed:
Remove console.logs
